lib/externalGraderCommon.js

```diff
@@ -117,11 +117,16 @@ module.exports.buildDirectory = function(dir, submission, variant, question, cou
                     // the directory hierarchy
                     const fullPath = path.join(dir, 'student', file.name);
                     const relPath = path.relative(path.join(dir, 'student'), fullPath);
-                    if (relPath.split(path.sep)[0] == '..' || path.isAbsolute(relPath)) {
+                    if (relPath.split(path.sep).includes('..') || path.isAbsolute(relPath)) {
                         callback(new Error('Invalid filename'));
                         return;
                     }
-                    fs.writeFile(path.join(dir, 'student', file.name), decodedContents, callback);
+                    // The version of ensureDir provided by fs-extra
+                    // automatically creates intermediate paths recursively.
+                    const targetDir = path.dirname(fullPath);
+                    fs.ensureDir(targetDir).then(() => {
+                        fs.writeFile(fullPath, decodedContents, callback);
+                    });
```

Comment:
The part that was in question. Is this version all right?

Reply:
Because we are in a callback-style function, we should be using old-style callback code to make sure no exceptions are thrown. This should be something like:
```
fs.ensureDir(targetDir, (err) => {
    if (ERR(err, callback)) return;
    fs.writeFile(fullPath, decodedContents, (err) => {
        if (ERR(err, callback)) return;
        callback(null);
    });
});
```

Reply:
See https://prairielearn.readthedocs.io/en/latest/dev-guide/#stack-traces-with-callback-style-functions for more details and examples of this.

Reply:
@mwest1066 I replaced the code with your suggestion. (I wonder if the innermost callback could have just been `callback` as before, though?)